routing/router.go

```diff
@@ -1668,119 +1743,209 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
 
 	timeoutChan := time.After(payAttemptTimeout)
 
+	paymentHash := payment.PaymentHash
+
 	// We'll continue until either our payment succeeds, or we encounter a
 	// critical error during path finding.
-	var lastError error
-	for {
+	var (
+		lastError error
+	)
+
+	// sendNewAttempt is a helper method that creates and sends a new
+	// payment to the switch. We use this when no payment has been sent to
+	// the switch already, or the previous attempt failed.
+	sendNewAttempt := func() error {
 		// Before we attempt this next payment, we'll check to see if
 		// either we've gone past the payment attempt timeout, or the
 		// router is exiting. In either case, we'll stop this payment
 		// attempt short.
 		select {
 		case <-timeoutChan:
+			// Mark the payment as failed.
+			err := a.Fail()
+			if err != nil {
+				return err
+			}
+
 			errStr := fmt.Sprintf("payment attempt not completed "+
 				"before timeout of %v", payAttemptTimeout)
 
-			return [32]byte{}, nil, newErr(
-				ErrPaymentAttemptTimeout, errStr,
-			)
+			// Terminal state, return.
+			return newErr(ErrPaymentAttemptTimeout, errStr)
 
 		case <-r.quit:
-			return [32]byte{}, nil, fmt.Errorf("router shutting down")
+			// The payment will be resumed from the current state
+			// after restart.
+			return ErrRouterShuttingDown
```

Comment:
Should this be `newErr` too?

Reply:
We are not acting on this error in any other way than other critical errors atm, so don't think it matters.